qemu/tests/cpu_device_hotplug.py

```diff
@@ -1,100 +1,119 @@
-import logging
 import re
 import time
+import random
+import logging
 
+from aexpect import ShellCmdError
 from virttest import error_context
 from virttest import utils_misc
-from virttest import cpu
+from virttest import cpu as utils_cpu
 
 
 @error_context.context_aware
 def run(test, params, env):
     """
-    Runs vCPU hotplug tests based on CPU device:
+    Test hot plug/unplug of vcpu device.
+
+    1) Boot up guest w/o vcpu device.
+    2) Hot plug/unplug vcpu device and check successfully or not. (qemu side)
+    3) Check if the number of CPUs in guest changes accordingly. (guest side)
+    4) Do sub test after hot plug/unplug.
+    5) Recheck the number of CPUs in guest.
+    6) Check the CPU topology of guest.
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
     """
 
-    def hotplug(vm, current_cpus, total_cpus, vcpu_threads):
-        for cpu in range(current_cpus, total_cpus):
-            error_context.context("hot-pluging vCPU %s" % cpu, logging.info)
-            vm.hotplug_vcpu(cpu_id=cpu, plug_command=hotplug_cmd)
-            time.sleep(0.1)
-        time.sleep(5)
-
-    def hotunplug(vm, current_cpus, total_cpus, vcpu_threads):
-        for cpu in range(current_cpus, total_cpus):
-            error_context.context("hot-unpluging vCPU %s" % cpu, logging.info)
-            vm.hotplug_vcpu(cpu_id=cpu, plug_command=unplug_cmd, unplug="yes")
-            time.sleep(0.1)
-        # Need more time to unplug, so sleeping more than hotplug.
-        time.sleep(10)
-
-    def verify(vm, total_cpus):
-        output = vm.monitor.send_args_cmd("info cpus")
-        logging.debug("Output of info CPUs:\n%s", output)
-
-        cpu_regexp = re.compile(r"CPU #(\d+)")
-        total_cpus_monitor = len(cpu_regexp.findall(output))
-        if total_cpus_monitor != total_cpus:
-            test.fail("Monitor reports %s CPUs, when VM should have"
-                      " %s" % (total_cpus_monitor, total_cpus))
-        error_context.context("hotplugging finished, let's wait a few sec and"
-                              " check CPUs quantity in guest.", logging.info)
-        if not utils_misc.wait_for(lambda: cpu.check_if_vm_vcpu_match(
-                                   total_cpus, vm),
-                                   60 + total_cpus, first=10,
-                                   step=5.0, text="retry later"):
-            test.fail("CPU quantity mismatch cmd after hotplug !")
-        error_context.context("rebooting the vm and check CPU quantity !",
+    def get_vm_cpu_ids():
```

Comment:
How about changing the function name to get_guest_cpu_ids since it's got inside guest.

Reply:
updated